File: src/assistant/gateway/schemas/acp.py
"""ACP listeners: install-level Connections (``platform="acp"``, ADR 0031) with a
live server task per configured port. Mirrors web/src/schemas/acp.ts.

A listener's shared token never rides a list or detail body (ADR 0024 spirit) —
only ``has_token`` says whether one is set. The raw value is returned exactly
once, from the create and rotate-token responses, and nowhere else.
"""

from pydantic import BaseModel


class AcpListenerOut(BaseModel):
    """One ACP listener as Settings shows it.

    ``running`` and ``error`` follow the same honesty pattern as a channel
    Connection's ``active``/``error`` (see ``ConnectionOut``): a listener that is
    configured but not running, or unreachable because its port collided, must not
    read the same as a healthy one.
    """

    id: str
    name: str
    profile: str
    port: int | None
    running: bool
    error: str | None
    has_token: bool


class AcpListenerListResponse(BaseModel):
    """GET /api/acp/listeners."""

    listeners: list[AcpListenerOut]


class AcpListenerCreatedResponse(BaseModel):
    """POST /api/acp/listeners. ``token`` is the raw shared secret — generated
    here when the request omits one — shown exactly this once so the owner can
    hand it to the client (AG2 Space); it is never echoed again. Empty for a
    stdio listener (no ``port``), which has no upgrade request to carry one."""

    listener: AcpListenerOut
    token: str


class AcpListenerTokenRotatedResponse(BaseModel):
    """POST /api/acp/listeners/{id}/rotate-token: the same one-time reveal as
    creation, since rotating is minting a new secret under the same id."""

    listener: AcpListenerOut
    token: str

File: web/src/schemas/acp.ts
// ACP listeners: install-level, one listener bound to one Profile at creation
// (ADR 0031) — never exposure-gated, so this file carries no exposure/default
// shapes the way connection.ts does for messaging platforms.
import { z } from 'zod'

// gateway/routes/acp.py _entry. `has_token` never carries the raw value — the
// secret is answered exactly once, from the create and rotate-token responses.
export const AcpListener = z.object({
  id: z.string(),
  name: z.string(),
  profile: z.string(),
  port: z.number().nullable(),
  running: z.boolean(),
  error: z.string().nullable(),
  has_token: z.boolean(),
})
export type AcpListener = z.infer<typeof AcpListener>

export const AcpListenerList = z.object({ listeners: z.array(AcpListener) })
export type AcpListenerList = z.infer<typeof AcpListenerList>

// POST /api/acp/listeners: the token is shown here once — Settings must copy it
// now or lose it forever. Empty for a stdio listener, which has none.
export const AcpListenerCreated = z.object({ listener: AcpListener, token: z.string() })
export type AcpListenerCreated = z.infer<typeof AcpListenerCreated>

// POST /api/acp/listeners/{id}/rotate-token: same one-time-reveal shape as create.
export const AcpListenerTokenRotated = z.object({ listener: AcpListener, token: z.string() })
export type AcpListenerTokenRotated = z.infer<typeof AcpListenerTokenRotated>
